patches/api/0419-Paper-Plugins.patch

```diff
@@ -0,0 +1,2123 @@
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/classloader/ClassLoaderAccess.java
+@@ -0,0 +1,10 @@
++package io.papermc.paper.plugin.provider.classloader;
++
++import org.jetbrains.annotations.ApiStatus;
++
++@ApiStatus.Internal
++public interface ClassLoaderAccess {
++
++    boolean canAccess(ConfiguredPluginClassLoader classLoader);
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/classloader/ConfiguredPluginClassLoader.java b/src/main/java/io/papermc/paper/plugin/provider/classloader/ConfiguredPluginClassLoader.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d86a88a97611092a270aa2f4004cab00780805c8
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/classloader/ConfiguredPluginClassLoader.java
+@@ -0,0 +1,19 @@
++package io.papermc.paper.plugin.provider.classloader;
++
++import io.papermc.paper.plugin.configuration.PluginMeta;
++import org.bukkit.plugin.java.JavaPlugin;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.io.Closeable;
++
++@ApiStatus.Internal
++public interface ConfiguredPluginClassLoader extends Closeable {
++
++    PluginMeta getConfiguration();
++
++    Class<?> loadClass(@NotNull String name, boolean resolve, boolean checkGlobal, boolean checkLibraries) throws ClassNotFoundException;
++
++    // Called in the constructor, at the very top
++    void init(JavaPlugin plugin);
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/classloader/PaperClassLoaderStorage.java b/src/main/java/io/papermc/paper/plugin/provider/classloader/PaperClassLoaderStorage.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..cb258b058e508967bffc4ef5a82bfcc2eee17e67
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/classloader/PaperClassLoaderStorage.java
+@@ -0,0 +1,30 @@
++package io.papermc.paper.plugin.provider.classloader;
++
++import net.kyori.adventure.util.Services;
++import org.bukkit.plugin.java.PluginClassLoader;
++import org.jetbrains.annotations.ApiStatus;
++
++@ApiStatus.Internal
++public interface PaperClassLoaderStorage {
++
++    static PaperClassLoaderStorage instance() {
++        return Access.INSTANCE;
++    }
++
++    PluginClassLoaderGroup registerSpigotGroup(PluginClassLoader pluginClassLoader);
++
++    PluginClassLoaderGroup registerOpenGroup(ConfiguredPluginClassLoader classLoader);
++
++    PluginClassLoaderGroup registerAccessBackedGroup(ConfiguredPluginClassLoader classLoader, ClassLoaderAccess access);
++
++    void unregisterClassloader(ConfiguredPluginClassLoader configuredPluginClassLoader);
++
++    boolean registerUnsafePlugin(ConfiguredPluginClassLoader pluginLoader);
++
++
++    @ApiStatus.Internal
++    class Access {
++
++        private static final PaperClassLoaderStorage INSTANCE = Services.service(PaperClassLoaderStorage.class).orElseThrow();
++    }
++}
```

Comment:
I don't remember the outcome of whatever discussion we had over this, but generally, extra package private final classes > useless public classes in api